app/black_spots/tasks/generate_training_input.py

```diff
@@ -0,0 +1,361 @@
+
+
+@shared_task
+def get_training_input(segments_shp_uuid, records_csv_uuid):
+    """Generate training input
+
+    Output to training csv to django-storages
+
+    Args:
+    :param segments_tar_uuid: (str) UUID for the RoadSegmentsShapefile to use
+    :param records_csv_uuid: (str) UUID for the BlackSpotRecordsFile
+
+    Returns:
+    TrainingCsvs: (namedtuple ['blackspots', 'forecast'])  tuple containing the
+                  BlackSpotTrainingCsv and ForecastTrainingCsv uuids
+    """
+
+    # Use passed uuid
+    with BlackSpotRecordsFile.objects.get(uuid=records_csv_uuid).csv as records_csv:
+        records, min_occurred, max_occurred = read_records(
+            records_csv, {'init': ROAD_PROJECTION}, {'init': RECORD_PROJECTION}, TIMEZONE,
+            RECORD_COL_ID, RECORD_COL_X, RECORD_COL_Y,
+            RECORD_COL_OCCURRED, RECORD_COL_SEVERE, RECORD_COL_SEVERE_VALS,
+            RECORD_COL_PRECIP, RECORD_COL_PRECIP_VALS
+        )
+    records_with_precip = [record for record in records if record['precip']]
+    records_without_precip = [record for record in records if not record['precip']]
+
+    # segments shapefile
+    tar_output_dir = tempfile.mkdtemp()
+    try:
+        shp_tar = RoadSegmentsShapefile.objects.get(uuid=segments_shp_uuid).shp_tgz.path
+        with tarfile.open(shp_tar, "r:gz") as tar:
+            tar.extractall(tar_output_dir)
+        with fiona.open(os.path.join(tar_output_dir, 'segments', COMBINED_SEGMENTS_SHP_NAME)) as segments:
+            segments_index = rtree.index.Index()
+            # Create a spatial index for segments to efficiently find nearby records
+            seg_shapes = [shape(segment['geometry']) for segment in segments]
+
+        for idx, element in enumerate(seg_shapes):
+            segments_index.insert(idx, element.bounds)
+
+        segments_with_records = match_records_to_segments(
+            records, segments_index, seg_shapes, MATCH_TOLERANCE
+        )
+        logger.info('Found {:,} segments with records'.format(len(segments_with_records)))
+
+        segments_with_records_precip = match_records_to_segments(
+            records_with_precip, segments_index, seg_shapes, MATCH_TOLERANCE
+        )
+        logger.info('Found {:,} segments with records w/precip'.format(len(segments_with_records_precip)))
+
+        segments_with_records_no_precip = match_records_to_segments(
+            records_without_precip, segments_index, seg_shapes, MATCH_TOLERANCE)
+
+        schema, segments_with_data = get_segments_with_data(
+            seg_shapes, segments_with_records, min_occurred, max_occurred
+        )
+        _, segments_with_data_precip = get_segments_with_data(
+            seg_shapes, segments_with_records_precip, min_occurred, max_occurred
+        )
+        _, segments_with_data_no_precip = get_segments_with_data(
+            seg_shapes, segments_with_records_no_precip, min_occurred, max_occurred
+        )
+    except:
```

Comment:
If I do that, it'll just throw another exception down the road on line 116 since `segments_with_data` won't be defined. Better to have it raise the initial exception for debugging purposes.